### ISSUE TITLE
Add tips for what should be enviroment variables in project.mds

### ISSUE THREAD

**Issue by astroash** [foundersandcoders/master-reference#803]
I noticed students often forgot what should be environment variables during projects. I think it could be useful in weeks 4-8 to have a **secret stuff**, which gives tips on what they should keep as environment variables

**Comment by shiryz** [comment]
@astroash this is a really good idea!

**Comment by eliascodes** [comment]
Where is the notion of env vars introduced? All I could find is a research topic in week 4. Is there anywhere else?

**Comment by astroash** [comment]
@eliascodes that's the only time its formally mentioned

**Comment by matthewdking** [comment]
I couldn't spot anything but it should be mentioned in week 3 as most teams end up using API keys in their projects

**Comment by eliascodes** [comment]
@matthewdking concealing API keys is mentioned in the project description for week 3. Anything beyond that is probably not useful because there would be no backend with which to conceal them.

**Comment by matthewdking** [comment]
@eliascodes good point :+1:

**Comment by astroash** [comment]
Yeah, I tried to get FAC12 to git ignore their api key and deploy on surge, just to keep it off of github, but it was a bit too much for them at the time.

**Comment by m4v15** [comment]
@eliascodes I'll need to check, but think it comes up during the code-along/walkthrough in database week as part of learn pg

**Comment by shannonjensen** [comment]
Hey all,

@tundeaji1232 and I are going through this for week4 planning right now. We agree the process for keeping environment variables secure isn't explained super well and there is room for improvement, but we have some questions about why it is done the way it is done. In my own personal experience, my teams always cribbed our db_connection.js files from various weeks' workshops (which changed week8 with use of pg-promise) and it wasn't clear how env2 worked.

In my last project, I used a different process, adapted from a Stephen Grider React course which I think might be more intuitive for students. Maybe we write up a little guide to environment variables and provide two ways of using them. That might make it easier for students to understand what is happening? I think using env2 and config.env instead of just a standard gitignored js file before databases (i.e. just for API key) might make things overly complicated?

The non-env2 way was to have three files in a config folder:

_config/keys.js_
```
if (process.env.NODE_ENV === "production") {
  module.exports = require("./prod");
} else {
  module.exports = require("./dev");
}
```

_config/prod.js_
```
module.exports = {
    DATABASE_URL: process.env.DATABASE_URL,
    COOKIE_KEY: process.env.COOKIE_KEY
    (any other secret keys)
};
```

_config/dev.js_
```
module.exports = {
  DATABASE_URL: "postgres://admin:password@localhost:5432/facevents",
  COOKIE_KEY: "fifughruioervhberisnfpcw;oq943qjrt894yt9p843h"
  (any other secret keys)
};
```

Then the db_connection.js file looks like this:
```
const pgp = require("pg-promise")();
const keys = require("../config/keys");
const url = require("url");

const params = url.parse(keys.DATABASE_URL);
const [username, password] = params.auth.split(":");

const connection = {
    host: params.hostname,
    database: params.pathname.split("/")[1],
    port: params.port
};

if (username) {
    connection.user = username;
}
if (password) {
    connection.password = password;
}

connection.ssl = connection.host !== "localhost";

const db = pgp(connection);
module.exports = db;
```

**Comment by shannonjensen** [comment]
Here is a guide @tundeaji1232 and I wrote up. Please let us know your thoughts.
https://hackmd.io/KYVghgZlDMDGC01gDYzwCwEYDsJ4E4pN4QRgAjSbaAJmVkyA

**Comment by eliascodes** [comment]
@shannonjensen thanks for this. I'm not familiar with the React course you mention, but I would say that the two approaches you outline are complimentary, rather than alternatives. In particular, I'd ask how you plan on deploying an app that exclusively uses the second approach.

Typically the way I use these approaches is to keep my _secrets_ in my environment variables (anything which must not be in source control), and use one or more `config.js` files to store my environment-dependent configuration. The latter will include both the secrets from the environment variables and other configuration stuff that doesn't need to be secret.
